deployment/tfhub/object-detection/url/object-detection-url-predict.ipynb

```diff
@@ -1,228 +1,341 @@
+      "outputs": [],
+      "source": [
+        "req = requests.get(url)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "5uN2lPRV9IyB"
+      },
+      "outputs": [],
+      "source": [
+        "soup = BeautifulSoup(req.text, 'lxml')"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "gH4BVlH_9I6g"
+      },
+      "outputs": [],
+      "source": [
+        "urls = []\n",
+        "\n",
+        "for link in soup.find_all('a'):\n",
+        "    href = link.get('href')\n",
+        "  \n",
+        "    if href.endswith('.jpg'):\n",
+        "      url = f\"http://www.cs.ucf.edu/~aroshan/index_files/Dataset_PitOrlManh/images/{href}\"\n",
+        "      urls.append(url)"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "nTIgnAsR7-OX"
+      },
+      "source": [
+        "## Show Process Metrics"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "BeqejBmJ73eE"
+      },
+      "outputs": [],
+      "source": [
+        "def show_metrics(urls, start_time, end_time):\n",
+        "    total_time = end_time - start_time\n",
+        "    total_time = time.strftime('%Hh %Mm %Ss', time.gmtime(total_time))\n",
+        "    \n",
+        "    print(f\"Processing Time: {total_time} for {len(urls)} URLs.\")"
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "llnSmebt8kMY"
+      },
+      "source": [
+        "## Predict"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "5wOeo8Q9_GIM"
+      },
+      "outputs": [],
+      "source": [
+        "def process_image(data):\n",
+        "    return model.predict(data)"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "0BuVkQo-9TTC"
+      },
+      "outputs": [],
+      "source": [
+        "# urls = urls[:100]\n",
+        "# urls = urls[:1000]\n",
+        "urls = urls[:10000]"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "Dk6Vg_6pAnGu"
+      },
+      "outputs": [],
+      "source": [
+        "data = [[os.path.basename(url), url] for url in urls]"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/"
+        },
+        "id": "zRGrVu36xaHU",
+        "outputId": "02380b02-f7df-4c11-8884-8af4ec6f6cd2"
+      },
+      "outputs": [],
+      "source": [
+        "results = []\n",
+        "start_time = time.time()\n",
+        "\n",
+        "with concurrent.futures.ThreadPoolExecutor() as executor:\n",
+        "    for result in zip(data, executor.map(process_image, data)):\n",
```

Comment:
I'm not personally too familiar with `concurrent.futures`! Is the `zip(data` required? Would it be possible to just use

```suggestion
        "    for result in executor.map(process_image, data):\n",
```

?

Reply:
With `zip()`, the `result` list will be filled with tuples containing the URL and the result dict of each image:
```
('http://www.cs.ucf.edu/~aroshan/index_files/Dataset_PitOrlManh/images/000001_0.jpg',
 {'file': '000001_0.jpg',
  'has_car': True,
  'score': 0.2524052858352661,
  'bboxes': {'ymin': 0.6500716209411621,
   'xmin': 0.08435962349176407,
   'ymax': 0.7936611175537109,
   'xmax': 0.28179624676704407}})
```

Without, `result` will be filled just with the result dict:
```
{'file': '000001_0.jpg',
 'has_car': True,
 'score': 0.2524052858352661,
 'bboxes': {'ymin': 0.6500716209411621,
  'xmin': 0.08435962349176407,
  'ymax': 0.7936611175537109,
  'xmax': 0.28179624676704407}}
```

As I wanted to return the result dict with the image file name, but it's already being done inside the `DetectObject` class, it's not necessary to use the `zip()` function here.

Great! :+1: